contents/javascript/Javascript-framework-VueJs.md

```diff
@@ -151,18 +153,105 @@ VueJs is a relatively new JavaScript framework as compared to Angular and React.
+```js
+this.message = 'Hello Space';
+```
+
+In React, such a direct modification of data is not allowed, due to React's need to rerun lifecycle hooks after state is being updated. Data can only be updated using `this.setState` method.
+```js
+this.setState({ message: 'Hello Space' });
+```
+<br>
+
+2. **2-way binding**: `v-model` is used to bind the DOM input field to its data variable. This effectively allows your DOM variables to be automatically in sync with your data, regardless of which one is being updated.
+In other words, this reduces the need for you to manually update your data.
+```js
+<input type="checkbox", v-model=isChecked">
+    <label for="checked">Select</label>
+```
+<br>
+
+3. **1-way data flow**: Data can only be passed from parent to child, via `props`.
+Props can be of any data type, including Objects.
+
+In the code below, `to-do list` is the parent and `item` is the child.
+The data in `item` is being passed to `todo-list` for rendering.
+```js
+Vue.component('todo-list', {
+    props: ['item'],
+    data: ['totalCount'],
+    template:
+      <div class='todo-list'>
+        <p>Total:{{ this.totalCount }}</p>
+        <p>{{ item.name }}: {{ item.count }}</p>
+})
+
+<todo-list
+  v-for='item in items'
+  v-bind:key='item.id'
+  v-bind:item='item'
+></todo-list>
+```
+
+
+However, what if the user decides to update the `item.count`? The data for item.count has to be passed from `item` to `todo-list` so `totalCount` can be updated inside `todo-list` . Under situations like this where the child has to pass data back to the parent, the child component has to [emit events](https://vuejs.org/v2/guide/components.html#Emitting-a-Value-With-an-Event)
+and the parent component will update after listening to these events.
+
+```js
+Vue.component('item', {
+  data: ['count', 'name'],
+  template: {
+    <button v-on:click="$emit('increased-count', count+1)">Increment count for this item</button>
+  }
+}
+
+// Inside todo-list component
+template:
+    v-on:increased-count="updateCount"
+```
+Whenever the button is pressed, an event called `increased-count` with the new value of `count` will be emitted by the child `item`.
+When `todo-list` listened to the event, it will execute `updateCount`.
+
+Using 1-way data flow ensures that the data can only be changed by the component itself and allows bugs to be easily traced in the code.
+<br>
+
+
+4. **Computed properties**: Useful when you want to reduce the amount of logic written in templates
+
+Using the example from above, we can convert `totalCount` into a computed property.
+
+```js
+computed: totalCount() {
+    let result = 0
+    this.items.forEach((item) => result += item.count);
+    return result;
+}
+```
+
+Unlike the use of methods, this updating of `totalCount` will only be triggered when the number of `items` in the list or any `item`'s `count` changed.
```

Comment:
There is a difference between the way you call the computed variables too, like with brackets and without. One will run the actual thing, one will use the cache, which is dependent on the state of the data that is used in the last "calculation". 

Its in the docs somewhere, like forcing the computed to always "compute". Could include for completeness sake

Reply:
Do you mean watched properties?